File: sdk/chocobase-csharp/Postgrest.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace ChocoBase
{
    public class QueryBuilder
    {
        private readonly string _baseUrl;
        private readonly string _table;
        private readonly HttpClient _httpClient;
        private readonly Dictionary<string, string> _headers;
        private readonly Dictionary<string, string> _params = new Dictionary<string, string>();

        public QueryBuilder(string baseUrl, string table, HttpClient httpClient, Dictionary<string, string> headers)
        {
            _baseUrl = baseUrl;
            _table = table;
            _httpClient = httpClient;
            _headers = headers;
        }

        public QueryBuilder Select(string columns = "*")
        {
            _params["select"] = columns;
            return this;
        }

        public QueryBuilder Eq(string column, object value)
        {
            _params[column] = $"eq.{value}";
            return this;
        }

        public QueryBuilder Limit(int count)
        {
            _params["limit"] = count.ToString();
            return this;
        }

        public async Task<List<Dictionary<string, object>>> ExecuteAsync()
        {
            var query = new StringBuilder();
            bool first = true;
            foreach (var kv in _params)
            {
                query.Append(first ? "?" : "&");
                query.Append(Uri.EscapeDataString(kv.Key));
                query.Append("=");
                query.Append(Uri.EscapeDataString(kv.Value));
                first = false;
            }

            var url = $"{_baseUrl}/rest/v1/{_table}{query}";
            using var req = new HttpRequestMessage(HttpMethod.Get, url);
            foreach (var h in _headers) req.Headers.TryAddWithoutValidation(h.Key, h.Value);

            var resp = await _httpClient.SendAsync(req);
            var json = await resp.Content.ReadAsStringAsync();

            try
            {
                return JsonSerializer.Deserialize<List<Dictionary<string, object>>>(json) ?? new List<Dictionary<string, object>>();
            }
            catch
            {
                return new List<Dictionary<string, object>>();
            }
        }
    }
}

File: sdk/chocobase-rb/lib/chocobase/postgrest.rb
module ChocoBase
  class Postgrest
    def initialize(base_url, table, headers)
      @base_url = base_url
      @table = table
      @headers = headers
      @params = {}
    end

    def select(columns = "*")
      @params["select"] = columns
      self
    end

    def eq(column, value)
      @params[column] = "eq.#{value}"
      self
    end

    def limit(count)
      @params["limit"] = count.to_s
      self
    end

    def execute
      uri = URI("#{@base_url}/rest/v1/#{@table}")
      uri.query = URI.encode_www_form(@params) unless @params.empty?

      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = (uri.scheme == "https")

      req = Net::HTTP::Get.new(uri.request_uri)
      @headers.each { |k, v| req[k] = v }

      res = http.request(req)
      parsed = JSON.parse(res.body) rescue []
      parsed.is_a?(Hash) && parsed.key?("rows") ? parsed["rows"] : parsed
    end
  end
end
